app/Libraries/UsernameValidation.php

```diff
@@ -82,33 +85,42 @@ public static function validateUsername($username)
     public static function validateUsersOfUsername(string $username): ValidationErrors
     {
         $errors = new ValidationErrors('user');
+        $userIds = static::usersOfUsername($username)->pluck('user_id');
 
-        $users = static::usersOfUsername($username);
-
-        // top 100
-        // Queried directly on model because User::rankHighests is disabled
-        // when experimental rank is set as default.
-        $highestRank = RankHighest::whereIn('user_id', $users->pluck('user_id'))->min('rank');
+        // Check if any of the users have been ranked in the top 100
+        $highestRank = RankHighest::whereIn('user_id', $userIds)->min('rank');
         if ($highestRank !== null && $highestRank <= 100) {
             return $errors->add('username', '.username_locked');
         }
 
-        foreach ($users as $user) {
-            // has badges
-            if ($user->badges()->exists()) {
-                return $errors->add('username', '.username_locked');
-            }
+        // Check if any of the users have badges
+        if (UserBadge::whereIn('user_id', $userIds)->exists()) {
+            return $errors->add('username', '.username_locked');
+        }
 
-            // ranked beatmaps
-            if ($user->profileBeatmapsetsRanked()->exists()) {
-                return $errors->add('username', '.username_locked');
-            }
+        // Check if any of the users are a host or guest of any Ranked,
+        // Approved, Qualified, or Loved beatmapsets
+        if (
+            Beatmapset
+                ::active()
+                ->withStates(['approved', 'loved', 'qualified', 'ranked'])
+                ->where(function (Builder $query) use ($userIds) {
+                    $query
+                        ->whereIn('user_id', $userIds)
+                        ->orWhereHas(
+                            'beatmaps',
+                            fn (Builder $query) => $query->whereIn('user_id', $userIds),
+                        );
+                })
+                ->exists()
```

Comment:
Checking beatmap state itself should be sufficient? Should usernames of pending guest diff that's part of loved map made available?

```
Beatmap::scoreable()->whereIn('user_id', $userIds)->exists()
```

Reply:
good point. fixed, although I have it checking beatmapset still because there are cases where the host doesn't mark themselves as owner of any map (common on collabs from what i can tell)